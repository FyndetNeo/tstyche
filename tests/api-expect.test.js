import { expect, test } from "@jest/globals";
import * as tstyche from "tstyche";
import { getFixtureUrl } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const fixtureUrl = getFixtureUrl("api-expect");

// TODO check for validation errors
// TODO currently 'expect()' cannot be nested because run mode flags are not inherited

test("handles '--failFast' option", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["fail-fast-expect.tst.ts", "--failFast"]);

  expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
  expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

  expect(exitCode).toBe(1);
});

test("'expect()' implementation'", () => {
  expect(tstyche.expect).toBeInstanceOf(Function);
});

test("'expect.fail' implementation'", () => {
  expect(tstyche.expect.fail).toBeInstanceOf(Function);
});

test("expect.fail", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["expect-fail.tst.ts"]);

  expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
  expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

  expect(exitCode).toBe(1);
});

test("'expect.only' implementation'", () => {
  expect(tstyche.expect.only).toBeInstanceOf(Function);
});

test("expect.only", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["expect-only.tst.ts"]);

  expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
  expect(stderr).toBe("");

  expect(exitCode).toBe(0);
});

test("'expect.only.fail' implementation'", () => {
  expect(tstyche.expect.only.fail).toBeInstanceOf(Function);
});

test("expect.only.fail", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["expect-only-fail.tst.ts"]);

  expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
  expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

  expect(exitCode).toBe(1);
});

test("'expect.skip' implementation'", () => {
  expect(tstyche.expect.skip).toBeInstanceOf(Function);
});

test("expect.skip", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["expect-skip.tst.ts"]);

  expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
  expect(stderr).toBe("");

  expect(exitCode).toBe(0);
});

test("'expect.skip.fail' implementation'", () => {
  expect(tstyche.expect.skip.fail).toBeInstanceOf(Function);
});

test("expect.skip.fail", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["expect-skip-fail.tst.ts"]);

  expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
  expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

  expect(exitCode).toBe(1);
});
