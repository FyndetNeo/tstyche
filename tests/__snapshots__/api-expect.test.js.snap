// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect.fail: stderr 1`] = `
"Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

  2 | 
  3 | expect<string>().type.toBeString();
> 4 | expect.fail<string>().type.toBeString();
    | ^
  5 | 
  6 | expect.fail<never>().type.toBeString();
  7 | 

      at ./__typetests__/expect-fail.tst.ts:4:1

Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

   9 |   test("is string?", () => {
  10 |     expect<string>().type.toBeString();
> 11 |     expect.fail<string>().type.toBeString();
     |     ^
  12 | 
  13 |     expect.fail<never>().type.toBeVoid();
  14 |   });

       at ./__typetests__/expect-fail.tst.ts:11:5

Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

  17 | test("is number?", () => {
  18 |   expect<number>().type.toBeNumber();
> 19 |   expect.fail<number>().type.toBeNumber();
     |   ^
  20 | 
  21 |   expect.fail<never>().type.toBeVoid();
  22 | });

       at ./__typetests__/expect-fail.tst.ts:19:3

"
`;

exports[`expect.fail: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

fail ./__typetests__/expect-fail.tst.ts
  is describe?
    × is string?
  × is number?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      2 failed, 2 total
Assertions: 3 failed, 6 passed, 9 total
Duration:   <<timestamp>>

Ran test files matching 'expect-fail.tst.ts'.
"
`;

exports[`expect.only.fail: stderr 1`] = `
"Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

  2 | 
  3 | expect<string>().type.toBeString();
> 4 | expect.only.fail<string>().type.toBeString();
    | ^
  5 | 
  6 | expect.only.fail<never>().type.toBeString();
  7 | 

      at ./__typetests__/expect-only-fail.tst.ts:4:1

Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

   9 |   test("is skipped?", () => {
  10 |     expect<never>().type.toBeVoid();
> 11 |     expect.only.fail<string>().type.toBeString();
     |     ^
  12 | 
  13 |     expect.only.fail<never>().type.toBeVoid();
  14 |   });

       at ./__typetests__/expect-only-fail.tst.ts:11:5

Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

  23 | test.only("is number?", () => {
  24 |   expect.skip<string>().type.toBeNumber();
> 25 |   expect.fail<number>().type.toBeNumber();
     |   ^
  26 | 
  27 |   expect.fail<never>().type.toBeVoid();
  28 | });

       at ./__typetests__/expect-only-fail.tst.ts:25:3

"
`;

exports[`expect.only.fail: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

fail ./__typetests__/expect-only-fail.tst.ts
  is skipped describe?
    - skip is skipped?
  is skipped describe?
    - skip is skipped?
  × is number?
  - skip is skipped?
  - todo is todo?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 3 skipped, 1 todo, 5 total
Assertions: 3 failed, 5 skipped, 3 passed, 11 total
Duration:   <<timestamp>>

Ran test files matching 'expect-only-fail.tst.ts'.
"
`;

exports[`expect.only: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/expect-only.tst.ts
  is describe?
    - skip is skipped?
  is skipped describe?
    - skip is skipped?
  + is number?
  - skip is skipped?
  - todo is todo?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      3 skipped, 1 todo, 1 passed, 5 total
Assertions: 5 skipped, 3 passed, 8 total
Duration:   <<timestamp>>

Ran test files matching 'expect-only.tst.ts'.
"
`;

exports[`expect.skip.fail: stderr 1`] = `
"Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

  2 | 
  3 | expect.skip.fail<string>().type.toBeString();
> 4 | expect.fail<string>().type.toBeString();
    | ^
  5 | 
  6 | describe("is describe?", () => {
  7 |   test("is string?", () => {

      at ./__typetests__/expect-skip-fail.tst.ts:4:1

Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

   7 |   test("is string?", () => {
   8 |     expect.skip.fail<string>().type.toBeString();
>  9 |     expect.fail<string>().type.toBeString();
     |     ^
  10 |   });
  11 | });
  12 | 

       at ./__typetests__/expect-skip-fail.tst.ts:9:5

Error: The assertion was supposed to fail, but it passed.

Consider removing the '.fail' flag.

  13 | test("is number?", () => {
  14 |   expect.skip.fail<number>().type.toBeNumber();
> 15 |   expect.fail<number>().type.toBeNumber();
     |   ^
  16 | });
  17 | 

       at ./__typetests__/expect-skip-fail.tst.ts:15:3

"
`;

exports[`expect.skip.fail: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

fail ./__typetests__/expect-skip-fail.tst.ts
  is describe?
    × is string?
  × is number?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      2 failed, 2 total
Assertions: 3 failed, 3 skipped, 6 total
Duration:   <<timestamp>>

Ran test files matching 'expect-skip-fail.tst.ts'.
"
`;

exports[`expect.skip: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/expect-skip.tst.ts
  is describe?
    + is string?
  is skipped describe?
    - skip is skipped?
  + is number?
  + is assignable?
  - skip is skipped?
  - todo is todo?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 todo, 3 passed, 6 total
Assertions: 6 skipped, 3 passed, 9 total
Duration:   <<timestamp>>

Ran test files matching 'expect-skip.tst.ts'.
"
`;

exports[`handles '--failFast' option: stderr 1`] = `
"Error: The source type is 'void'.

   5 | });
   6 | 
>  7 | expect<void>().type.toBeUndefined();
     |                     ^
   8 | 
   9 | describe("is describe?", () => {
  10 |   test("is number?", () => {

       at ./__typetests__/fail-fast-expect.tst.ts:7:21

"
`;

exports[`handles '--failFast' option: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

fail ./__typetests__/fail-fast-expect.tst.ts
  + is string?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 failed, 1 passed, 2 total
Duration:   <<timestamp>>

Ran test files matching 'fail-fast-expect.tst.ts'.
"
`;
